### ISSUE TITLE
Multithreading issues

### ISSUE THREAD

**Issue by markbaird** [pencilblue/news-sitemap#3]
I'm getting some really bizarre and inconsistent results out of this plugin.

Problem 1: The TemplateService instance is being reused. Multiple threads are setting values on the single template service instance at the same time. This is causing the results to look like:
`[Article 2, Article 2, Article 4, Article 4, Article 6, Article 6]`
instead of the expected:
`[Article 1, Article 2, Article 3, Article 4, Article 5, Article 6]`

There is another issue where articles are missing from the results. I've noticed an issue where the context seems to change in the middle of the calls to ArticleService.getMetaInfo(). If it gets called simultaneously with "Article 1" and "Article 2", then you can print each article correctly to the logs at the beginning of that function, but at the end of the function both process will have instances of "Article 2". Whenever this happens, "Article 1" will be missing from the results. I would expect there to be duplicate instances of "Article 2" in the results when this happens, but that isn't occurring.

These issues go away if I set PARALLEL_LIMIT=1 to prevent multiple threads, but that's obviously not the best solution.

**Comment by brianhyder** [comment]
You are correct.  Would you mind trying something for us?  Can you move line 76 of *controllers/news_sitemap.js* to line 81?  I think that will solve the issue. 

Should look something like:
```
    NewsSitemap.prototype.processObjects = function(objArray, cb) {
        var self = this;

        var tasks = util.getTasks(objArray, function(objArray, i) {
            return function(callback) {
                ArticleService.getMetaInfo(objArray[i], function(metaKeywords, metaDescription, metaTitle) {
                    var ts   = new pb.TemplateService(self.ls);
                    ts.registerLocal('url', '/article/' + objArray[i].url);
                    ts.registerLocal('publish_date', self.getPublishDate(objArray[i].publish_date));
                    ts.registerLocal('headline', metaTitle);
                    ts.registerLocal('keywords', metaKeywords);
                    ts.load('xml_feeds/news_sitemap/url', callback);
                });
            };
        });
        async.parallelLimit(tasks, PARALLEL_LIMIT, function(err, results) {
            cb(err, results.join(''));
        });
    };
```

**Comment by markbaird** [comment]
I actually tried making that change already. I was going to submit it as a pull request. However it only fixes the first issue I listed.

The other issue is more difficult to figure out. The query will return 5 articles, there will be 5 async tasks created, and I will only get 4 articles in the results. When I add lots of logging to see what is going on it becomes obvious that there is some sort of thread issue where in the middle of an ArticleService.getMetaInfo() call the value of the article will change. I can't pin down what is causing it though.

**Comment by brianhyder** [comment]
I got a chance to look at this one this evening.  I think I have a fix for it.  I'll tag you on the bit in the PR where I made the change.  Great catch BTW.  I'll also be pushing the fix to this plugin once we get green lights from Travis and merge.

**Comment by markbaird** [comment]
Thanks for looking at it! I'm sorry I couldn't help more but I just couldn't figure out what was going in inside the getMetaInfo() function when called in parallel. I'll test the fix tomorrow to verify it fixes the issue I was seeing.

**Comment by brianhyder** [comment]
No worries.  Let me know if you have any other issues with it.

**Comment by markbaird** [comment]
After pulling down the latest version of PencilBlue and updating to the latest version of this plugin all issues appear to be fixed.

Thanks again.

**Comment by brianhyder** [comment]
Awesome. Thank you for confirming.

**Comment by markbaird** [comment]
I think I tested it without properly updating the plugin. After going through the step of updating the git submodule in order to push it to OpenShift I'm seeing a bug in the plugin. I will fix it and send a pull request.
